tests/lava/proc/dense/test_learning.py

```diff
@@ -0,0 +1,1649 @@
+        num_steps : int
+            Number of simulation time steps.
+        weights_init : ndarray
+            Initial weights matrix for the LearningDense.
+        learning_rule_cnd : str
+            String specifying which learning rule condition to use.
+        graded_spike_cfg : GradedSpikeCfg
+            GradedSpikeCfg to use for the LearningDense.
+
+        Returns
+        ----------
+        pattern_pre : RingBuffer
+            Pre-synaptic RingBuffer Process.
+        learning_dense : LearningDense
+            LearningDense Process.
+        pattern_post : RingBuffer
+            Post-synaptic RingBuffer Process.
+        """
+
+        if learning_rule_cnd == "x" or learning_rule_cnd == "y":
+            scaling_exp = -1
+        elif learning_rule_cnd == "u":
+            scaling_exp = -2
+        else:
+            scaling_exp = 0
+
+        dw = f"2^{scaling_exp} * {learning_rule_cnd}0 * x1"
+
+        x1_impulse = 16 if graded_spike_cfg == GradedSpikeCfg.DEFAULT else 0
+
+        learning_rule = Loihi2FLearningRule(dw=dw,
+                                            x1_impulse=x1_impulse, x1_tau=12,
+                                            x2_impulse=24, x2_tau=5,
+                                            y1_impulse=16, y1_tau=10,
+                                            y2_impulse=24, y2_tau=5,
+                                            t_epoch=4, rng_seed=0)
+
+        spike_raster_pre = np.zeros((1, num_steps))
+        spike_raster_pre[0, 4] = 51
+        spike_raster_pre[0, 11] = 100
+        spike_raster_pre[0, 17] = 60
+        spike_raster_pre[0, 21] = 246
+        spike_raster_pre[0, 38] = 30
+        spike_raster_pre[0, 55] = 22
+        spike_raster_post = np.zeros((1, num_steps))
+        spike_raster_post[0, 3] = 1
+        spike_raster_post[0, 21] = 1
+        spike_raster_post[0, 39] = 1
+        spike_raster_post[0, 57] = 1
+
+        pattern_pre = RingBuffer(data=spike_raster_pre.astype(int))
+        learning_dense = \
+            LearningDense(weights=weights_init,
+                          learning_rule=learning_rule,
+                          name="learning_dense",
+                          num_message_bits=8,
+                          graded_spike_cfg=graded_spike_cfg)
+        pattern_post = RingBuffer(data=spike_raster_post.astype(int))
+
+        pattern_pre.s_out.connect(learning_dense.s_in)
+        pattern_post.s_out.connect(learning_dense.s_in_bap)
+
+        return pattern_pre, learning_dense, pattern_post
+
+    expected_x1_data_default = \
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 12.0, 12.0, 12.0, 12.0, 24.0, 24.0,
+         24.0, 24.0, 19.0, 19.0, 19.0, 19.0, 27.0, 27.0, 27.0, 27.0, 34.0, 34.0,
+         34.0, 34.0, 25.0, 25.0, 25.0, 25.0, 19.0, 19.0, 19.0, 19.0, 15.0, 15.0,
+         15.0, 15.0, 26.0, 26.0, 26.0, 26.0, 20.0, 20.0, 20.0, 20.0, 16.0, 16.0,
+         16.0, 16.0, 12.0, 12.0, 12.0, 12.0, 24.0, 24.0, 24.0, 24.0, 18.0, 18.0,
+         18.0, 18.0]
```

Comment:
Where do these values come from? How do you know that these are expected? Can we make that more explicit so that the test code becomes more readable to people?

Do we need 63 time steps to make sure the algorithm is working as expected? If it were just 5, for instance, the difference between the different modes would be more obvious - I guess.

Reply:
These values come from the simulation itself. They have been thoroughly validated (in the sense that they approximate the chip behavior). The purpose of these known value tests is to help us notice breaking changes in the future (which would make these tests fail.

As for the second question, the choice of 63 is arbitrary here. It could have been something else. It is true that the fact that number is high makes the difference between the different modes less obvious to the user/developer reading this code snippet.

However, the results we are testing for here depend not only on the chosen mode, but also on:
1. The learning rule condition (for each mode, all three different learning rule conditions are tested)
2. The epoch length (set to 4 in all these tests)
3. Pre-synaptic spiking history
4. Pre-synaptic within-epoch spike times
5. Post-synaptic spiking history (in some cases)

The purpose of having a high enough number of time steps is for making the pre/post-synaptic spiking history and pre-synaptic within epoch spike times more "diverse", so that we can cover more cases in each test.

Reply:
I was hoping we could design simple tests that cover all the features and are still understandable to people. Then you could actually write out the expected numbers by hand, explain why you expect them in the test, and anyone reading the tests later on could have a chance at understanding why those numbers are expected.

Currently, if the test suddenly fails, the only thing a developer can understand is that the values are different. He or she cannot see why they should be what you expect unless they go through all the learning engine code in detail and understand all of that.

If the tests depend on these five factors you are listing, maybe there should really be a test for each of these factors in isolation.

Reply:
Makes perfect sense. Addressed this in a new commit.